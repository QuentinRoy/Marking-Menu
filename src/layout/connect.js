import { finalize, tap } from 'rxjs/operators';
import rafThrottle from 'raf-throttle';

/**
 * Connect navigation notifications to menu opening and closing.
 *
 * @param {HTMLElement} parentDOM - The element where to append the menu.
 * @param {Observable} navigation$ - Notifications of the navigation.
 * @param {Function} createMenuLayout - Menu layout factory.
 * @param {Function} createUpperStrokeCanvas - Upper stroke canvas factory. The
 * upper stroke show the user interaction on the current menu, and the movements
 * in expert mode.
 * @param {Function} createLowerStrokeCanvas - Lower stroke canvas factory. The
 * lower stroke is stroke drawn below the menu. It keeps track of the previous
 * movements.
 * @param {Function} createGestureFeedback - Create gesture feedback.
 * @param {{error}} log - Logger.
 * @return {Observable} `navigation$` with menu opening and closing side effects.
 */
export default (
  parentDOM,
  navigation$,
  createMenuLayout,
  createUpperStrokeCanvas,
  createLowerStrokeCanvas,
  createGestureFeedback,
  log
) => {
  // The menu object.
  let menu = null;
  // A stroke drawn on top of the menu.
  let upperStrokeCanvas = null;
  // A stroke drawn below the menu.
  let lowerStrokeCanvas = null;
  // The points of the lower strokes.
  let lowerStroke = null;
  // The points of the upper stroke.
  let upperStroke = null;

  const gestureFeedback = createGestureFeedback(parentDOM);

  const closeMenu = () => {
    menu.remove();
    menu = null;
  };

  const openMenu = (model, position) => {
    const cbr = parentDOM.getBoundingClientRect();
    menu = createMenuLayout(parentDOM, model, [
      position[0] - cbr.left,
      position[1] - cbr.top
    ]);
  };

  const setActive = id => {
    menu.setActive(id);
  };

  const startUpperStroke = position => {
    upperStrokeCanvas = createUpperStrokeCanvas(parentDOM);
    upperStroke = [position];
    upperStrokeCanvas.drawStroke(upperStroke);
  };

  const noviceMove = rafThrottle(position => {
    if (upperStrokeCanvas) {
      upperStrokeCanvas.clear();
      if (position) {
        upperStroke = [upperStroke[0], position];
        upperStrokeCanvas.drawStroke(upperStroke);
      }
      upperStrokeCanvas.drawPoint(upperStroke[0]);
    }
  });

  const expertDraw = rafThrottle(stroke => {
    // FIXME: Not very efficient.
    if (upperStrokeCanvas) {
      upperStrokeCanvas.clear();
      upperStroke = stroke.slice();
      upperStrokeCanvas.drawStroke(upperStroke);
    }
  });

  const clearUpperStroke = () => {
    upperStrokeCanvas.remove();
    upperStrokeCanvas = null;
    upperStroke = null;
  };

  const swapUpperStroke = () => {
    lowerStroke = lowerStroke ? [...lowerStroke, ...upperStroke] : upperStroke;
    clearUpperStroke();
    lowerStrokeCanvas = lowerStrokeCanvas || createLowerStrokeCanvas(parentDOM);
    lowerStrokeCanvas.drawStroke(lowerStroke);
  };

  const clearLowerStroke = () => {
    if (lowerStrokeCanvas) {
      lowerStrokeCanvas.remove();
      lowerStrokeCanvas = null;
    }
    lowerStroke = null;
  };

  const showGestureFeedback = isCanceled => {
    gestureFeedback.show(
      lowerStroke ? [...lowerStroke, ...upperStroke] : upperStroke,
      isCanceled
    );
  };

  const cleanUp = () => {
    // Make sure everything is cleaned upon completion.
    if (menu) closeMenu();
    if (upperStrokeCanvas) clearUpperStroke();
    if (lowerStrokeCanvas) clearLowerStroke();
    gestureFeedback.remove();
    // eslint-disable-next-line no-param-reassign
    parentDOM.style.cursor = '';
  };

  const onNotification = notification => {
    switch (notification.type) {
      case 'open': {
        // eslint-disable-next-line no-param-reassign
        parentDOM.style.cursor = 'none';
        if (menu) closeMenu();
        swapUpperStroke();
        openMenu(notification.menu, notification.center);
        startUpperStroke(notification.center);
        noviceMove(notification.position);
        break;
      }
      case 'change': {
        setActive((notification.active && notification.active.id) || null);
        break;
      }
      case 'select':
      case 'cancel':
        // eslint-disable-next-line no-param-reassign
        parentDOM.style.cursor = '';
        if (menu) closeMenu();
        showGestureFeedback(notification.type === 'cancel');
        clearUpperStroke();
        clearLowerStroke();
        break;
      case 'start':
        // eslint-disable-next-line no-param-reassign
        parentDOM.style.cursor = 'crosshair';
        startUpperStroke(notification.position);
        break;
      case 'draw':
        expertDraw(notification.stroke);
        break;
      case 'move':
        noviceMove(notification.position);
        break;
      default:
        throw new Error(
          `Invalid navigation notification type: ${notification.type}`
        );
    }
  };

  return navigation$.pipe(
    tap({
      next(notification) {
        try {
          onNotification(notification);
        } catch (e) {
          log.error(e);
          throw e;
        }
      },
      error(e) {
        log.error(e);
        throw e;
      }
    }),
    finalize(() => {
      try {
        cleanUp();
      } catch (e) {
        log.error(e);
        throw e;
      }
    })
  );
};
