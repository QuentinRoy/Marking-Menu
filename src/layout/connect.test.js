import { of } from 'rxjs';
import { marbles } from 'rxjs-marbles';
import connect from './connect';

jest.mock('raf-throttle', () => jest.fn(f => f));

// Mock values.
let UpperStrokeCanvas;
let LowerStrokeCanvas;
let strokeCanvasInstances;
let MenuLayout;
let menuLayoutInstances;
let parentElement;
let log;

beforeEach(() => {
  parentElement = document.createElement('div');

  strokeCanvasInstances = [];
  const createStrokeCanvasFactory = name =>
    jest.fn(parent => {
      const div = document.createElement('div');
      div.className = `${name}-stroke-canvas`;
      parent.appendChild(div);

      // Set up the strokes properties as data attributes.
      div.dataset.points = JSON.stringify([]);
      div.dataset.stroke = null;

      const self = {
        name,
        mock: { div, parent, removed: false },
        drawStroke: jest.fn(stroke => {
          div.dataset.stroke = JSON.stringify(stroke);
        }),
        drawPoint: jest.fn(p => {
          div.dataset.points = JSON.stringify([
            ...JSON.parse(div.dataset.points),
            p
          ]);
        }),
        clear: jest.fn(() => {
          div.dataset.stroke = null;
          div.dataset.points = JSON.stringify([]);
        }),
        remove: jest.fn(() => {
          div.remove();
          self.mock.removed = true;
        })
      };
      strokeCanvasInstances.push(self);
      return self;
    });
  UpperStrokeCanvas = createStrokeCanvasFactory('upper');
  LowerStrokeCanvas = createStrokeCanvasFactory('lower');

  menuLayoutInstances = [];
  MenuLayout = jest.fn((parent, model, center, active) => {
    const div = document.createElement('div');
    div.className = 'menu';
    parent.appendChild(div);

    // Set up the menu properties as data attributes.
    div.dataset.model = JSON.stringify(model);
    div.dataset.center = JSON.stringify(center);
    div.dataset.active = JSON.stringify(active);

    const self = {
      mock: { div, parent, removed: false, active: null },
      setActive: jest.fn(newActive => {
        div.dataset.active = JSON.stringify(newActive);
      }),
      remove: jest.fn(() => {
        div.remove();
        self.mock.removed = true;
      })
    };
    menuLayoutInstances.push(self);
    return self;
  });

  log = {
    error: jest.fn()
  };
});

describe('connect', () => {
  it('draws expert strokes on draw notifications', done => {
    const src = of(
      { type: 'start', position: 'pos1' },
      { type: 'draw', stroke: 'stroke1' },
      { type: 'draw', stroke: 'stroke2' },
      { type: 'select' },
      { type: 'start', position: 'pos2' },
      { type: 'draw', stroke: 'stroke3' },
      { type: 'cancel' },
      { type: 'start', position: 'pos3' }
    );

    const out = connect(
      parentElement,
      src,
      MenuLayout,
      UpperStrokeCanvas,
      LowerStrokeCanvas
    );
    out.subscribe({
      next() {
        expect(parentElement).toMatchSnapshot();
      },
      complete() {
        expect(parentElement).toMatchSnapshot();
        done();
      }
    });
  });

  it('opens the menu, draws novice stroke, and updates the menu', done => {
    parentElement.getBoundingClientRect = jest.fn(() => ({ left: 1, top: 2 }));
    const src = of(
      { type: 'start', position: 'pos1' },
      { type: 'draw', stroke: [[0, 0], [1, 1]] },
      { type: 'open', menu: 'menu1', center: [10, 20], position: [100, 200] },
      { type: 'move', position: [1000, 2000] },
      { type: 'change', active: { id: 'active-item-1' } },
      { type: 'move', position: [500, 700] },
      { type: 'move', position: [100, 200] },
      { type: 'change', active: undefined },
      { type: 'change', active: { id: 'active-item-2' } },
      { type: 'select' },

      { type: 'start', position: 'pos2' },
      { type: 'draw', stroke: [[1, 1], [0, 0]] },
      { type: 'open', menu: 'menu2', center: [30, 40], position: [300, 800] },
      { type: 'move', position: [200, 900] },
      { type: 'change', active: { id: 'active-item-3' } },
      { type: 'open', menu: 'menu3', center: [60, 70], position: [300, 800] },
      { type: 'change', active: { id: 'active-item-4' } },
      { type: 'move', position: [700, 100] },
      { type: 'cancel' },

      { type: 'start', position: 'pos3' },
      { type: 'open', menu: 'menu4', center: [90, 30], position: [102, 201] },
      { type: 'move', position: [500, 0] },
      { type: 'change', active: { id: 'active-item-5' } }
    );

    const out = connect(
      parentElement,
      src,
      MenuLayout,
      UpperStrokeCanvas,
      LowerStrokeCanvas
    );
    out.subscribe({
      next() {
        expect(parentElement).toMatchSnapshot();
      },
      complete() {
        expect(parentElement).toMatchSnapshot();
        done();
      }
    });
  });

  // prettier-ignore
  it('cleans-up on error', marbles(m => {
    const values = {
      s: { type: 'start', position: 'pos1' },
      o: {
        type: 'open',
        menu: 'menu1',
        center: [10, 20],
        position: [100, 200]
      }
    };
    const error = new Error('Test Error');
    const obs = m.hot('---s--o--#', values, error);
    const exp = m.hot('---s--o--#', values, error);
    m.expect(
      connect(
        parentElement,
        obs,
        MenuLayout,
        UpperStrokeCanvas,
        LowerStrokeCanvas,
        log
      )
    ).toBeObservable(exp);
    m.flush();
    expect(parentElement).toMatchSnapshot();
    expect(log.error.mock.calls).toEqual([[error]]);
  }));

  // prettier-ignore
  it('cleans-up and throws on unknown event type', marbles(m => {
    const values = {
      s: { type: 'start', position: 'pos1' },
      o: {
        type: 'open',
        menu: 'menu1',
        center: [10, 20],
        position: [100, 200]
      },
      x: { type: 'unknown' }
    };
    const error = new Error('Invalid navigation notification type: unknown');
    const obs = m.hot('---s--o--x-', values);
    const exp = m.hot('---s--o--#', values, error);
    m.expect(
      connect(
        parentElement,
        obs,
        MenuLayout,
        UpperStrokeCanvas,
        LowerStrokeCanvas,
        log
      )
    ).toBeObservable(exp);
    m.flush();
    expect(parentElement).toMatchSnapshot();
    expect(log.error.mock.calls).toEqual([[error]]);
  }));
});
