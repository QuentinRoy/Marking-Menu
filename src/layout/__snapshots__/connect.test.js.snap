// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`connect cleans-up and throws on unknown event type 1`] = `
<div
  style=""
/>
`;

exports[`connect cleans-up on error 1`] = `
<div
  style=""
/>
`;

exports[`connect draws expert strokes on draw notifications 1`] = `
<div
  style="cursor: crosshair;"
>
  <div
    class="upper-stroke-canvas"
    data-points="[]"
    data-stroke="[\\"pos1\\"]"
  />
</div>
`;

exports[`connect draws expert strokes on draw notifications 2`] = `
<div
  style="cursor: crosshair;"
>
  <div
    class="upper-stroke-canvas"
    data-points="[]"
    data-stroke="\\"stroke1\\""
  />
</div>
`;

exports[`connect draws expert strokes on draw notifications 3`] = `
<div
  style="cursor: crosshair;"
>
  <div
    class="upper-stroke-canvas"
    data-points="[]"
    data-stroke="\\"stroke2\\""
  />
</div>
`;

exports[`connect draws expert strokes on draw notifications 4`] = `
<div
  style=""
/>
`;

exports[`connect draws expert strokes on draw notifications 5`] = `
<div
  style="cursor: crosshair;"
>
  <div
    class="upper-stroke-canvas"
    data-points="[]"
    data-stroke="[\\"pos2\\"]"
  />
</div>
`;

exports[`connect draws expert strokes on draw notifications 6`] = `
<div
  style="cursor: crosshair;"
>
  <div
    class="upper-stroke-canvas"
    data-points="[]"
    data-stroke="\\"stroke3\\""
  />
</div>
`;

exports[`connect draws expert strokes on draw notifications 7`] = `
<div
  style=""
/>
`;

exports[`connect draws expert strokes on draw notifications 8`] = `
<div
  style="cursor: crosshair;"
>
  <div
    class="upper-stroke-canvas"
    data-points="[]"
    data-stroke="[\\"pos3\\"]"
  />
</div>
`;

exports[`connect draws expert strokes on draw notifications 9`] = `
<div
  style="cursor: crosshair;"
>
  <div
    class="upper-stroke-canvas"
    data-points="[]"
    data-stroke="[\\"pos3\\"]"
  />
</div>
`;

exports[`connect opens the menu, draws novice stroke, and updates the menu 1`] = `
<div
  style="cursor: crosshair;"
>
  <div
    class="upper-stroke-canvas"
    data-points="[]"
    data-stroke="[\\"pos1\\"]"
  />
</div>
`;

exports[`connect opens the menu, draws novice stroke, and updates the menu 2`] = `
<div
  style="cursor: crosshair;"
>
  <div
    class="upper-stroke-canvas"
    data-points="[]"
    data-stroke="[[0,0],[1,1]]"
  />
</div>
`;

exports[`connect opens the menu, draws novice stroke, and updates the menu 3`] = `
<div
  style="cursor: none;"
>
  <div
    class="lower-stroke-canvas"
    data-points="[]"
    data-stroke="[[0,0],[1,1]]"
  />
  <div
    class="menu"
    data-active="undefined"
    data-center="[9,18]"
    data-model="\\"menu1\\""
  />
  <div
    class="upper-stroke-canvas"
    data-points="[[10,20]]"
    data-stroke="[[10,20],[100,200]]"
  />
</div>
`;

exports[`connect opens the menu, draws novice stroke, and updates the menu 4`] = `
<div
  style="cursor: none;"
>
  <div
    class="lower-stroke-canvas"
    data-points="[]"
    data-stroke="[[0,0],[1,1]]"
  />
  <div
    class="menu"
    data-active="undefined"
    data-center="[9,18]"
    data-model="\\"menu1\\""
  />
  <div
    class="upper-stroke-canvas"
    data-points="[[10,20]]"
    data-stroke="[[10,20],[1000,2000]]"
  />
</div>
`;

exports[`connect opens the menu, draws novice stroke, and updates the menu 5`] = `
<div
  style="cursor: none;"
>
  <div
    class="lower-stroke-canvas"
    data-points="[]"
    data-stroke="[[0,0],[1,1]]"
  />
  <div
    class="menu"
    data-active="\\"active-item-1\\""
    data-center="[9,18]"
    data-model="\\"menu1\\""
  />
  <div
    class="upper-stroke-canvas"
    data-points="[[10,20]]"
    data-stroke="[[10,20],[1000,2000]]"
  />
</div>
`;

exports[`connect opens the menu, draws novice stroke, and updates the menu 6`] = `
<div
  style="cursor: none;"
>
  <div
    class="lower-stroke-canvas"
    data-points="[]"
    data-stroke="[[0,0],[1,1]]"
  />
  <div
    class="menu"
    data-active="\\"active-item-1\\""
    data-center="[9,18]"
    data-model="\\"menu1\\""
  />
  <div
    class="upper-stroke-canvas"
    data-points="[[10,20]]"
    data-stroke="[[10,20],[500,700]]"
  />
</div>
`;

exports[`connect opens the menu, draws novice stroke, and updates the menu 7`] = `
<div
  style="cursor: none;"
>
  <div
    class="lower-stroke-canvas"
    data-points="[]"
    data-stroke="[[0,0],[1,1]]"
  />
  <div
    class="menu"
    data-active="\\"active-item-1\\""
    data-center="[9,18]"
    data-model="\\"menu1\\""
  />
  <div
    class="upper-stroke-canvas"
    data-points="[[10,20]]"
    data-stroke="[[10,20],[100,200]]"
  />
</div>
`;

exports[`connect opens the menu, draws novice stroke, and updates the menu 8`] = `
<div
  style="cursor: none;"
>
  <div
    class="lower-stroke-canvas"
    data-points="[]"
    data-stroke="[[0,0],[1,1]]"
  />
  <div
    class="menu"
    data-active="null"
    data-center="[9,18]"
    data-model="\\"menu1\\""
  />
  <div
    class="upper-stroke-canvas"
    data-points="[[10,20]]"
    data-stroke="[[10,20],[100,200]]"
  />
</div>
`;

exports[`connect opens the menu, draws novice stroke, and updates the menu 9`] = `
<div
  style="cursor: none;"
>
  <div
    class="lower-stroke-canvas"
    data-points="[]"
    data-stroke="[[0,0],[1,1]]"
  />
  <div
    class="menu"
    data-active="\\"active-item-2\\""
    data-center="[9,18]"
    data-model="\\"menu1\\""
  />
  <div
    class="upper-stroke-canvas"
    data-points="[[10,20]]"
    data-stroke="[[10,20],[100,200]]"
  />
</div>
`;

exports[`connect opens the menu, draws novice stroke, and updates the menu 10`] = `
<div
  style=""
/>
`;

exports[`connect opens the menu, draws novice stroke, and updates the menu 11`] = `
<div
  style="cursor: crosshair;"
>
  <div
    class="upper-stroke-canvas"
    data-points="[]"
    data-stroke="[\\"pos2\\"]"
  />
</div>
`;

exports[`connect opens the menu, draws novice stroke, and updates the menu 12`] = `
<div
  style="cursor: crosshair;"
>
  <div
    class="upper-stroke-canvas"
    data-points="[]"
    data-stroke="[[1,1],[0,0]]"
  />
</div>
`;

exports[`connect opens the menu, draws novice stroke, and updates the menu 13`] = `
<div
  style="cursor: none;"
>
  <div
    class="lower-stroke-canvas"
    data-points="[]"
    data-stroke="[[1,1],[0,0]]"
  />
  <div
    class="menu"
    data-active="undefined"
    data-center="[29,38]"
    data-model="\\"menu2\\""
  />
  <div
    class="upper-stroke-canvas"
    data-points="[[30,40]]"
    data-stroke="[[30,40],[300,800]]"
  />
</div>
`;

exports[`connect opens the menu, draws novice stroke, and updates the menu 14`] = `
<div
  style="cursor: none;"
>
  <div
    class="lower-stroke-canvas"
    data-points="[]"
    data-stroke="[[1,1],[0,0]]"
  />
  <div
    class="menu"
    data-active="undefined"
    data-center="[29,38]"
    data-model="\\"menu2\\""
  />
  <div
    class="upper-stroke-canvas"
    data-points="[[30,40]]"
    data-stroke="[[30,40],[200,900]]"
  />
</div>
`;

exports[`connect opens the menu, draws novice stroke, and updates the menu 15`] = `
<div
  style="cursor: none;"
>
  <div
    class="lower-stroke-canvas"
    data-points="[]"
    data-stroke="[[1,1],[0,0]]"
  />
  <div
    class="menu"
    data-active="\\"active-item-3\\""
    data-center="[29,38]"
    data-model="\\"menu2\\""
  />
  <div
    class="upper-stroke-canvas"
    data-points="[[30,40]]"
    data-stroke="[[30,40],[200,900]]"
  />
</div>
`;

exports[`connect opens the menu, draws novice stroke, and updates the menu 16`] = `
<div
  style="cursor: none;"
>
  <div
    class="lower-stroke-canvas"
    data-points="[]"
    data-stroke="[[1,1],[0,0],[30,40],[200,900]]"
  />
  <div
    class="menu"
    data-active="undefined"
    data-center="[59,68]"
    data-model="\\"menu3\\""
  />
  <div
    class="upper-stroke-canvas"
    data-points="[[60,70]]"
    data-stroke="[[60,70],[300,800]]"
  />
</div>
`;

exports[`connect opens the menu, draws novice stroke, and updates the menu 17`] = `
<div
  style="cursor: none;"
>
  <div
    class="lower-stroke-canvas"
    data-points="[]"
    data-stroke="[[1,1],[0,0],[30,40],[200,900]]"
  />
  <div
    class="menu"
    data-active="\\"active-item-4\\""
    data-center="[59,68]"
    data-model="\\"menu3\\""
  />
  <div
    class="upper-stroke-canvas"
    data-points="[[60,70]]"
    data-stroke="[[60,70],[300,800]]"
  />
</div>
`;

exports[`connect opens the menu, draws novice stroke, and updates the menu 18`] = `
<div
  style="cursor: none;"
>
  <div
    class="lower-stroke-canvas"
    data-points="[]"
    data-stroke="[[1,1],[0,0],[30,40],[200,900]]"
  />
  <div
    class="menu"
    data-active="\\"active-item-4\\""
    data-center="[59,68]"
    data-model="\\"menu3\\""
  />
  <div
    class="upper-stroke-canvas"
    data-points="[[60,70]]"
    data-stroke="[[60,70],[700,100]]"
  />
</div>
`;

exports[`connect opens the menu, draws novice stroke, and updates the menu 19`] = `
<div
  style=""
/>
`;

exports[`connect opens the menu, draws novice stroke, and updates the menu 20`] = `
<div
  style="cursor: crosshair;"
>
  <div
    class="upper-stroke-canvas"
    data-points="[]"
    data-stroke="[\\"pos3\\"]"
  />
</div>
`;

exports[`connect opens the menu, draws novice stroke, and updates the menu 21`] = `
<div
  style="cursor: none;"
>
  <div
    class="lower-stroke-canvas"
    data-points="[]"
    data-stroke="[\\"pos3\\"]"
  />
  <div
    class="menu"
    data-active="undefined"
    data-center="[89,28]"
    data-model="\\"menu4\\""
  />
  <div
    class="upper-stroke-canvas"
    data-points="[[90,30]]"
    data-stroke="[[90,30],[102,201]]"
  />
</div>
`;

exports[`connect opens the menu, draws novice stroke, and updates the menu 22`] = `
<div
  style="cursor: none;"
>
  <div
    class="lower-stroke-canvas"
    data-points="[]"
    data-stroke="[\\"pos3\\"]"
  />
  <div
    class="menu"
    data-active="undefined"
    data-center="[89,28]"
    data-model="\\"menu4\\""
  />
  <div
    class="upper-stroke-canvas"
    data-points="[[90,30]]"
    data-stroke="[[90,30],[500,0]]"
  />
</div>
`;

exports[`connect opens the menu, draws novice stroke, and updates the menu 23`] = `
<div
  style="cursor: none;"
>
  <div
    class="lower-stroke-canvas"
    data-points="[]"
    data-stroke="[\\"pos3\\"]"
  />
  <div
    class="menu"
    data-active="\\"active-item-5\\""
    data-center="[89,28]"
    data-model="\\"menu4\\""
  />
  <div
    class="upper-stroke-canvas"
    data-points="[[90,30]]"
    data-stroke="[[90,30],[500,0]]"
  />
</div>
`;

exports[`connect opens the menu, draws novice stroke, and updates the menu 24`] = `
<div
  style="cursor: none;"
>
  <div
    class="lower-stroke-canvas"
    data-points="[]"
    data-stroke="[\\"pos3\\"]"
  />
  <div
    class="menu"
    data-active="\\"active-item-5\\""
    data-center="[89,28]"
    data-model="\\"menu4\\""
  />
  <div
    class="upper-stroke-canvas"
    data-points="[[90,30]]"
    data-stroke="[[90,30],[500,0]]"
  />
</div>
`;
